xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-filters/xwiki-platform-notifications-filters-api/src/main/java/org/xwiki/notifications/filters/internal/DefaultNotificationFilterPreferenceManager.java

```diff
@@ -56,55 +58,99 @@
 @Singleton
 public class DefaultNotificationFilterPreferenceManager implements NotificationFilterPreferenceManager
 {
-    private static final String ENABLE_ERROR_MESSAGE = "Failed to enable or disabled the filter preference [{}].";
-
     @Inject
     private ComponentManager componentManager;
 
     @Inject
     private Logger logger;
 
-    @Override
-    public Collection<NotificationFilterPreference> getFilterPreferences(DocumentReference user)
-            throws NotificationException
+    @FunctionalInterface
+    interface ProviderCallable
     {
-        Set<NotificationFilterPreference> filterPreferences = new HashSet<>();
-
-        try {
-            List<NotificationFilterPreferenceProvider> providers
-                    = componentManager.getInstanceList(NotificationFilterPreferenceProvider.class);
+        void doInProvider(NotificationFilterPreferenceProvider provider) throws NotificationException;
+    }
 
-            for (NotificationFilterPreferenceProvider provider : providers) {
-                filterPreferences.addAll(provider.getFilterPreferences(user));
-            }
+    @FunctionalInterface
+    interface RetrieveWithProviderCallable<E>
+    {
+        Collection<E> retrieveWithProvider(NotificationFilterPreferenceProvider provider)
+            throws NotificationException;
+    }
 
-            return filterPreferences;
+    private List<NotificationFilterPreferenceProvider> getProviderList() throws NotificationException
+    {
+        try {
+            return componentManager.getInstanceList(NotificationFilterPreferenceProvider.class);
         } catch (ComponentLookupException e) {
-            throw new NotificationException(
-                String.format("Unable to fetch a list of notification preference providers with user [%s].", user));
+            throw new NotificationException("Error when trying to load the list of providers", e);
         }
     }
 
-    @Override
-    public Collection<NotificationFilterPreference> getFilterPreferences(WikiReference wikiReference)
-        throws NotificationException
+    private String getProviderDebugMessage(String loggerMessage, NotificationFilterPreferenceProvider provider)
     {
-        Set<NotificationFilterPreference> filterPreferences = new HashSet<>();
+        return String.format("%s with provider %s", loggerMessage, provider);
+    }
 
-        try {
-            List<NotificationFilterPreferenceProvider> providers
-                = componentManager.getInstanceList(NotificationFilterPreferenceProvider.class);
+    private String getExceptionMessage(String loggerMessage, List<NotificationException> exceptions)
+    {
+        return String.format("%s - All providers called failed, see exceptions: [%s].",
```

Comment:
I'm honestly not sure this message is report is enough, now displaying all stacktraces might not be readable... I'm taking suggestion.

Reply:
I don't have a much better idea.